### PR TITLE
Not stopping recovery while still closing the opened channel

### PR DESCRIPTION
**Pull request by Inlustris**
## Proposed Changes

Started the change for cleaning up the connection while failing at recovery.  I don't know if this is what you are looking for but i hope it gets us started to create a correct solution. 

By creating the internal function and an overload for AutorecoveringConnection also internal. Nothing changes for the current users of the nuget.

## Types of Changes

- [x] Bug fix (non-breaking change which fixes issue #1067)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq) (already did for previous bug #1061)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (It is hard to test even if bug #1061 is completed because of that's the only reason i know why it was failing)
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

## Further Comments

No unit test is made because i don't know a situation it should fail in. The issue i had in #1061 resulted in the recover loop. 
With the next unit test i could recreate the issue (until #1061 is completed). The next unit test still uses the Model inside the recover objects which will be removed in #1061. 

Until the line with dashes you will be able to debug. And see a recover loop every 5 seconds. 

```
[Fact]
public void TestConsumerRecoveryOnClientWithDeclarationOnOneChannelAndConsumersOnTheSecondChannel()
{
    string q = "dotnet-client.recovery.queue1";
    string x = "tmp-fanout";
    string rk = "tmp-routingkey";

    using (AutorecoveringConnection c = CreateAutorecoveringConnection())
    {
        IModel m = c.CreateModel();

        m.ExchangeDeclare(exchange: x, type: "fanout");
        m.QueueDeclare(q, false, false, false, null);
        m.QueueBind(q, x, rk);
        m.Close();

        IModel m1 = c.CreateModel();

        var cons = new EventingBasicConsumer(m1);
        m1.BasicConsume(q, true, cons);
        AssertConsumerCount(m1, q, 1);

        CloseAndWaitForRecovery(c);
-------------------------------------------------------------------------------------
        AssertConsumerCount(m1, q, 1);

        m1.ExchangeDeclarePassive(x);
        m1.QueueDeclarePassive(q);

        var latch = new ManualResetEventSlim(false);
        cons.Received += (s, args) => latch.Set();

        m1.BasicPublish("", q, null, _encoding.GetBytes("msg"));
        Wait(latch);

        m1.QueueUnbind(q, x, rk);
        m1.ExchangeDelete(x);
        m1.QueueDelete(q);
    }
}

```

### CODE REVIEW

File: projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.Recovery.cs [hunk 1]
```diff
@@ -158,6 +161,11 @@ private bool TryPerformAutomaticRecovery()
             }
             catch (Exception e)
             {
+                if(connectionRecoverySucceeded)
+                {
+                    this.Close(506, "Recovery failed.", true);
```
Comment (michaelklishin) [inline projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.Recovery.cs:166]: Note that RabbitMQ won't do anything special about a `5xx` response (besides logging the error message).

Comment (Inlustris) [reply projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.Recovery.cs:166]: I looked up the code's in the AMQP spec and think this was the most matching one. Not a specific reason for it.

Comment (michaelklishin) [reply projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.Recovery.cs:166]: This is not particularly specific. Should it say "channel recovery failed"?

Comment (Inlustris) [reply projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.Recovery.cs:166]: No it is a general issue kicking in when exception is catched. At this point we don't know i think or we should look at the exception.

**Comment by michaelklishin** [comment]
So after the dashed line in the test recovery loop is stopped? can we somehow assert on it, e.g. by introducing a recovery state getter used by tests?

**Comment by Inlustris** [comment]
It will timeout on the test. But after the dashes you are not able to check consumers etc. because the connection is shutdown.

You mean a state within the code that the test can use or a state within the test?

File: projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.Recovery.cs [hunk 2]
```diff
@@ -158,6 +161,11 @@ private bool TryPerformAutomaticRecovery()
             }
             catch (Exception e)
             {
+                if(connectionRecoverySucceeded)
```
Comment (sungam3r) [inline projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.Recovery.cs:164]:
```suggestion
                if (connectionRecoverySucceeded)
```

**Comment by Inlustris** [comment]
@michaelklishin Same for this one? If anything needs to be done still let me know.

**Comment by michaelklishin** [comment]
@Inlustris can you please rebase this against `main`? Then I'd be happy to review one more time. Thanks.

**Comment by Inlustris** [comment]
I had some Rebase issues. But i think now it's correct. @michaelklishin

**Comment by Inlustris** [comment]
@lukebakken Can you confirm this is an other implementation of #1145.

And means this one can be closed?

**Comment by lukebakken** [comment]
I'll confirm when I have time, thanks.

**Comment by lukebakken** [comment]
> Can you confirm this is an other implementation of https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1145.

Yes! I will close this PR. Thank you for your contributions.

**Pull request closed without merging**